File: core/kora-app-annotation-processor/src/test/java/io/koraframework/kora/app/annotation/processor/app/AppWithFactories12.java
package io.koraframework.kora.app.annotation.processor.app;

import io.koraframework.common.annotation.KoraApp;
import io.koraframework.common.annotation.Root;

@KoraApp
public interface AppWithFactories12 {
    @Root
    default Object mock1(ConfigValueMapper<TestEnum> object) {
        return new Object();
    }

    default <T extends Enum<T>> EnumConfigValueMapper<T> enumConfigValueMapper() {
        return new EnumConfigValueMapper<>();
    }

    enum TestEnum {}

    interface ConfigValueMapper<T> {
        T extract(Object value);
    }

    class EnumConfigValueMapper<T extends Enum<T>> implements ConfigValueMapper<T> {
        @Override
        public T extract(Object value) {
            return null;
        }
    }

}

File: kora-app-symbol-processor/src/test/kotlin/io/koraframework/kora/app/ksp/app/AppWithFactories12.kt
package io.koraframework.kora.app.ksp.app

import io.koraframework.common.KoraApp
import io.koraframework.common.annotation.Root

@KoraApp
interface AppWithFactories12 {

    fun <T, R> factory1(t: SimpleGeneric<T>, r: SimpleGeneric<R>): GenericClass<T, R> {
        return GenericClass(t, r)
    }

    fun factory2(): SimpleGeneric<Long> {
        return SimpleGeneric()
    }

    fun factory3(): SimpleGeneric<String> {
        return SimpleGeneric()
    }

    @Root
    fun mock1(cl: GenericClass<Long, String>): Any = Any()

    open class SimpleGeneric<T>
    data class GenericClass<T, R>(val t: SimpleGeneric<T>, val r: SimpleGeneric<R>)
}
